invocation.md

```diff
@@ -0,0 +1,1170 @@
+    },
+    "sendTextMessage": {
+      "<-": [
+        {
+          "/": "bafybeibwlfwol5bdwj75hdqs3liv6z7dyqwtd2t4ovvgncv4ixkaxlkfle"
+        },
+        "ok"
+      ]
+    }
+  },
+  "s": {
+    "/": {
+      "bytes:": "5vNn4--uTeGk_vayyPuNTYJ71Yr2nWkc6AkTv1QPWSgetpsu8SHegWoDakPVTdxkWb6nhVKAz6JdpgnjABppC7"
+    }
+  }
+}
+```
+
+# 4 Receipt
+
+An Invocation Receipt is an attestation of the Result of an Invocation. A Receipt MUST be signed by the Executor (the `aud` of the associated UCANs) or it's delegate, in which case proof of delegation (of the invoked capability) from Executor to the Signer (the `iss` of the receipt) MUST be provided in `prf`.
+
+**NB: a Receipt this does not guarantee correctness of the result!** The statement's veracity MUST be only understood as an attestation from the executor.
+
+Receipts MUST use the same version as the invocation that they contain.
+
+## 4.1 Schema
+
+```ipldsch
+type Receipt<In, Out> struct {
+  job     &Invocation<In>
+
+  # output of the invocation
+  out     Out
+
+  # Related receipts
+  origin optional &Receipt<In, Any>
+
+  # All the other metadata
+  meta    { String: Any }
+
+  # Principal that issued this receipt
+  iss     Principal
+
+  # Proof that `iss` was authorized to by invocation `aud` to issue
+  # receipts. Can be omitted if `job.aud === this.iss`
+  prf     [&UCAN] implicit ([])
+
+  # Signature from the `iss`.
+  s       Varsig
+}
+```
+
+## 4.2 Fields
+
+### 4.2.1 Invocation
+
+The `job` field MUST include a link to the Invocation that the Receipt is for.
+
+### 4.2.2 Output
+
+The `out` field MUST contain the output of the invocation. It is RECOMMENDED that invocation produce output in [Result] format or it's extension with added type variants when needed, which allows selecting either success or failure branch during promise pipelining.
+
+### 4.2.2.1 Result
+
+A Result records success or a failure state of the [Invocation].
+
+#### 4.2.2.1.1 Variants
+
+```ipldsch
+type Result<T, X> union {
+  | T  ("ok")    # Success
+  | X  ("error") # Failure
+} representation keyed
+```
+
+#### 4.2.2.1.2 Success
+
+The success branch MUST contain the value returned from a successful invocation wrapped in the `"ok"` tag. The exact shape of the returned data is left undefined to allow for flexibility.
+
+```json
+{ "ok": 42 }
+```
+
+#### 4.2.2.1.3 Failure
+
+The failure branch MAY contain detail about why execution failed wrapped in the `"error"` tag. The exact shape of the returned data is left undefined to allow for flexibility.
+
+If no information is available, this field SHOULD be set to `{}`.
+
+```json
+{
+  "error": {
+    "dev/reason": "unauthorized",
+    "http/status": 401
+  }
+}
+```
+
+### 4.2.2.2 Extended Result
+
+An extension of Result records SHOULD be used when result of [Invocation] does not fit binary success / failure criteria.
+
+```ipldsch
+type Status<T, X, P> union {
+  | T  ("ok")      # Success
+  | X  ("error")   # Failure
+  | P  ("pending") # Pending
```

Comment:
Where would this be used? In an RPC call?

Reply:
We had a conversation about it other day where you've convinced me to just stick to ok / error. I have not updated PR to reflect this because my understanding was we wanted to align on group auth stuff first which is sketched in https://github.com/web3-storage/specs/pull/35/files